runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,340 @@
+          TransformConsumer.create(
+              responseFunction, StreamObserverConsumer.create(responseObserver));
+      invocation.addStateListener(stateListener);
+    } catch (Exception e) {
+      String errMessage =
+          String.format("Encountered Unexpected Exception for Invocation %s", invocationId);
+      LOG.error(errMessage, e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void getMessageStream(
+      JobMessagesRequest request,
+      StreamObserver<JobMessagesResponse> responseObserver) {
+    String invocationId = request.getJobId();
+    try {
+      JobInvocation invocation = getInvocation(invocationId);
+      // synchronization is necessary since we are multiplexing this stream observer.
+      responseObserver = SynchronizedStreamObserver.wrapping(responseObserver);
+
+      Function<JobState.Enum, JobMessagesResponse> stateResponseFunction =
+          state ->
+              JobMessagesResponse
+                  .newBuilder()
+                  .setStateResponse(GetJobStateResponse.newBuilder().setState(state).build())
+                  .build();
+      Consumer<JobState.Enum> stateListener =
+          TransformConsumer.create(
+              stateResponseFunction, StreamObserverConsumer.create(responseObserver));
+
+      Function<JobMessage, JobMessagesResponse> messagesResponseFunction =
+          message -> JobMessagesResponse.newBuilder().setMessageResponse(message).build();
+      Consumer<JobMessage> messageListener =
+          TransformConsumer.create(
+              messagesResponseFunction, StreamObserverConsumer.create(responseObserver));
+
+      invocation.addStateListener(stateListener);
+      invocation.addMessageListener(messageListener);
+    } catch (Exception e) {
+      String errMessage =
+          String.format("Encountered Unexpected Exception for Invocation %s", invocationId);
+      LOG.error(errMessage, e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (JobPreparation preparation : ImmutableList.copyOf(preparations.values())) {
+      try {
+        preparation.stagingService().close();
+      } catch (Exception e) {
+        LOG.warn("Exception while closing job {}", preparation);
+      }
+    }
+    // TODO: should we cancel managed jobs when this gets closed?
+  }
+
+  private JobInvocation getInvocation(String invocationId) throws StatusException {
+    JobInvocation invocation = invocations.get(invocationId);
+    if (invocation == null) {
+      throw Status.NOT_FOUND.asException();
+    }
+    return invocation;
+  }
+
+  /**
+   * Forward inputs to a StreamObserver.
+   */
+  private static class StreamObserverConsumer<T> implements Consumer<T> {
+    public static <T> StreamObserverConsumer<T> create(StreamObserver<T> sink) {
+      return new StreamObserverConsumer<>(sink);
```

Comment:
Seems that this can be implemented as "return sink::onNext", or perhaps this class can be removed altogether in favor of that.

Reply:
Good point.  That's a much better solution for both consumer classes.